### PR TITLE
feat: migrate to Assemblyscript

### PR DESCRIPTION
**Pull request by marvinroger**

### CODE REVIEW

File: packages/nanocurrency/src/assembly/utils.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+export function reverseBytes(input: Uint8Array): void {
+  let i = input.length - 1;
+  let j = 0;
+  while (i > j) {
+    let tmp = input[i];
+    input[i] = input[j];
+    input[j] = tmp;
+    i--;
+    j++;
+  }
```
Comment (MaxGraey) [inline packages/nanocurrency/src/assembly/utils.ts]: Hi, just use `input.reverse()`

File: packages/nanocurrency/src/assembly/blake2b.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  h: Uint64Array;
+  t: Uint64Array;
+  c: usize;
+  outlen: usize;
+}
+
+const HASH_LENGTH: u32 = 8;
+
+function ROTR64(x: u64, y: u64): u64 {
+  return (x >> y) ^ (x << (64 - y));
```
Comment (MaxGraey) [inline packages/nanocurrency/src/assembly/blake2b.ts]: AS has `rotr<u64>(x, y)` builtin

File: packages/nanocurrency/package.json [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  "module": "dist/nanocurrency.esm.js",
+  "types": "dist/types/index.d.ts",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/marvinroger/nanocurrency-js.git"
+  },
+  "scripts": {
+    "build:prod": "yarn build:prod:assembly && yarn build:prod:js",
+    "build:prod:js": "rimraf dist/ && cross-env NODE_ENV=production rollup -c",
+    "build:prod:assembly": "asc src/assembly/index.ts -b tmp/production.wasm --validate --optimize",
```
Comment (MaxGraey) [inline packages/nanocurrency/package.json]: For best performance better use `-O3` instead `--optimize`

File: packages/nanocurrency/src/assembly/utils.ts (hunk 1 above)
Comment (MaxGraey) [reply packages/nanocurrency/src/assembly/utils.ts]: Ahh, no it's already implemented but not in dist files yet

Comment (marvinroger) [reply packages/nanocurrency/src/assembly/utils.ts]: I saw that, but I was not able to use it when I worked on it.
I tried updating to the latest git rev, and it does work, so I guess it's in the dist?

Anyway, thanks for the review. This is a POC, an attempt to drop Emscripten, but I have a lot of areas to optimize yet (10x performance drop).

File: packages/nanocurrency/src/assembly/blake2b.ts (hunk 2 above)
Comment (marvinroger) [reply packages/nanocurrency/src/assembly/blake2b.ts]: Thanks!

File: packages/nanocurrency/src/assembly/utils.ts (hunk 1 above)
Comment (MaxGraey) [reply packages/nanocurrency/src/assembly/utils.ts]: Correct. It's not in the dist files yet. So will be allow in next release

**Comment by marvinroger** [comment]
Closing to open a new draft PR.

**Pull request closed without merging**